handler/crash_report_upload_thread.cc

```
@@ -264,6 +265,8 @@ void CrashReportUploadThread::ProcessPendingReport(
             time(nullptr) +
             (1 << upload_report->upload_attempts) * kRetryWorkIntervalSeconds;
       }
+#elif BUILDFLAG(IS_ANDROID)
+      upload_report.reset();
```

This implies the report will be retried indefinitely as there doesn't seem to be a check for kRetryAttempts (mentioned in the git log) in this case. Note the iOS block has `if (upload_report->upload_attempts > kRetryAttempts)`. Does this mean that a user's storage will be filled up with crash reports if they can never send them?

---

Thanks for the comments! they’re all hitting the same core issue so I’ll answer them together;

On Android today, Crashpad is skipping offline crashes (on transient failures and on rate‑limit) and those reports are never retried again. 
The files are still on disk, just marked `completed / skipped` instead of `pending`. 
For Backtrace’s Android integration we need Crashpad to be absolutely `reliable in offline / flaky‑network scenarios` so this change is deliberately moving Android away from silently skipping toward keep pending until it successfully uploads or an explicit retention policy deletes it. 
The Android backend (SDK) handles the retry behaviour and longer term cleanup / retention is a follow‑up not in scope of this PR.

---

A bit more detail:

- Infinite retries / storage

  Before this patch, if the device was offline or the URL misconfigured we would hit `UploadResult::kRetry` and immediately call `SkipReportUpload()`. 
That doesn’t delete the minidump it only flips its state to *completed / skipped* and guarantees it will never be retried. So a permanently broken config could already fill disk with crash reports (they just lived in the completed state instead of pending).

  On Android we’re choosing not to call `SkipReportUpload` on `kRetry` so we keep the same “one file per crash” behavior but preserve the ability to eventually send them once network/config is fixed. The Backtrace Android backend runs its own `CrashReportUploadThread` and disables Crashpad’s rate limit there so pending reports are flushed on the next online startup instead of sitting around indefinitely.

- Rate limiting / multiple pending reports / completed

  The existing rate‑limit `ShouldRateLimitUpload` was effectively a `skip` as well: if we hit the 1 hour window, the report was marked skipped and never retried. There’s no reaper for completed reports either so this already had “indefinite” behaviour but in the completed state.

  Now we treat this as “delay this attempt and keep the report pending” rather than “give up forever”. On Android, we don’t actually rely on this rate limiter in our app‑side uploader (we pass options tailored for that use case) and we want to avoid silently losing data.

- Erroneous time / orphaned reports

  Same story here: previously this code would `SkipReportUpload()` and strand the report in completed. That’s also an indefinite state and the report would never be uploaded. On Android, we’d rather keep it pending and give the system a chance to upload.

So yes! we’re intentionally changing Android semantics away from “skip” in these cases. That’s necessary to make Crashpad usable as a reliable offline store in the Android backend we’re plugging it into. The existing `// TODO` around retry semantics and rate limiting in Crashpad is a good signal that the generic policy here probably deserves a deeper redesign. We’d like to tackle proper retention / expiry as a separate piece of work instead of overloading `kRetry` handling in this PR.

Scope of this change is therefore:

- Fix Android’s offline reliability gap (don’t permanently skip reports).
- Keep behaviour for other platforms unchanged.
- Leave longer term policy questions (max attempts / retention / disk budget) to a follow‑up where we can design them explicitly rather than as side effects of `SkipReportUpload()`.